### ISSUE TITLE
Evaluation module boilerplate

### ISSUE THREAD

**Issue by aimalz** [LSSTDESC/rail_attic#4]
[The PZ DC1 metrics code](https://github.com/LSSTDESC/PZDC1paper/tree/master/metric_scripts) can be used as a starting point for the submodules that evaluate the accuracy of photo-z PDF estimates.  Note that many of those functions were built on [qp](https://github.com/aimalz/qp), which is at this point pretty old and could use an upgrade, which will occur in [a DESC fork](https://github.com/LSSTDESC/qp).

Steps (Julia):  (starting at sprint week 2020)

- [x] Clone the RAIL repository and create a new branch (follow the branch name convention).

- [x] Quick review of PZ DC1 paper.

- [x] Study the current implementation available in the PZ DC1 metrics code. 

- [x] Find toy data to run quick tests while coding. 

- [x] Run FlexZBoost on the toy data to get input PDFs. 

- [x] Create a superclass for the generic concept of metric. It will be the mother of a number of classes, one for each metric. It should receive PDFs from any photo-z algorithm (standardized format), and allows the comparison between results from several codes. 

- [x] Create the metric classes (start with one PDF metric, e.g. the CDE loss, and run tests before going to the full list of metrics and point estimate reduction). 
 

- [x] Create a demo notebook showing the metrics evaluation using the new classes.

- [x] Run using the exact same data used in the DC1 paper to allow the reproduction of the paper's results (numbers in tables 2 and 3 from the paper must match). 

- [ ]  Write unit tests for the new classes.
- [ ] Update repo's documentation file.

- [ ] Open a pull request.

**Comment by gschwend** [comment]
progress report: I had trouble to run FlexZBoost, on nersc, due to conda env issues. I am getting help from Johann Cohen-Tanugi via slack. We agreed that the best approach for now is to work locally in my laptop, which also is giving us a hard time to setup rail's dependencies (but we are almost there).

**Comment by gschwend** [comment]
Finally, I can FlexZBoost on the test sample successfully. 

The requirements to run on my laptop were: 
```
$ conda create -n py37 python=3.7 pyyaml pandas 
$ conda activate py37
$ pip install flexcode[all]
$ conda install -c conda-forge openmp
$ /bin/bash -c "$(curl -fsSL https://raw.githubusercontent.com/Homebrew/install/HEAD/install.sh)"
$ brew install libomp
 ``` 
Maybe it is a good idea to update the setup.py file to include all the dependencies.

**Comment by sschmidt23** [comment]
There was some discussion about how to manage package dependencies for individual codes, as it might get burdensome once we have many codes to have everyone install every package dependency when an individual user may only want to run one code.  As this discussion is ongoing, flexcode is not included in the install dependencies in the setup.py file.  I think the comment from @Gschwend may be evidence that we should just bite the bullet and do a full install_requires for necessary packages.

Note that you do not need to use condo, you can look at the GitHub actions setup to see the packages necessary to install rail and run all unit tests using pip:
https://github.com/LSSTDESC/RAIL/blob/master/.github/workflows/main.yml

**Comment by sschmidt23** [comment]
And FYI, on cori, I either make my own custom conda environment, or do a 
`module load python/3.7-anaconda-2019.07`
then install any necessary extra packages with e.g.
`python setup.py install --prefix=[path to local install location]`
and then add `[path to local install location]/lib/python3.7/site-packages` to my $PYTHONPATH.

**Comment by sschmidt23** [comment]
I'll create an issue to fix up the setup.py.  I think we can use extras_require and have different install requirements for each individual package, plus a "Full" for installing everything, which should be pretty straightforward (modulo the fact that xgboost is sometimes a pain to install)"

**Comment by gschwend** [comment]
progress report: I copied DC1 paper metrics code to a temporary directory here for refactoring (starting with the conversion from Python 2 to Python 3).

**Comment by gschwend** [comment]
progress report: I am in the process of writing the classes based on DC1 metrics code. One minor issue has been to deal with input data. The data used in qp example notebook has a companion file with the metadata. The results from codes I ran using  RAIL estimation module do not come with this metadata file. I am creating it by hand. No big deal, just to inform the progress...

**Comment by johannct** [comment]
can you copy over or attach these metadata, I do not remember the DC1 formats

**Comment by sschmidt23** [comment]
I don't remember a metadata file in the "old" qp, I think the updated qp creates a metadata file when writing out astropy tables, but those would not exist for the PZDC1 results.  For PZDC1 we asked users to submit their final PDFs as plain ASCII text columns.  There was some departures from exactly what we asked for in a few cases, I dealt with this with separate ingest functions for each code (it was messy but it worked at the time).  
The outputs for individual codes are currently in:
/global/cfs/cdirs/lsst/groups/PZ/PhotoZDC1/photoz_results/
in subdirectories for each code named `TESTDC1[codename]`
Each of those directories should have a `zarrayfile.out` that specifies which redshift grid was used in the `[codename]gold_pz.out` file containing the i<25.3 sample of 399,356 galaxies that we used in PZDC1.

We just did everything in ascii for PZDC1, so the read in of data will be different for your example data read in from RAIL where we use hdf5.  Neither should have a metadata file, though...

**Comment by eacharles** [comment]
FWIW, the new qp write the metadata file b/c not all of the formats from writing astropy table support writing multiple tables in a single file.

If we were to pick a particular format, just as hdf5 or fits, that does support it, we could dispense with the meta data table.   Or we could try and be clever and check.

-e

**Comment by gschwend** [comment]
Progress report: now I already have the results of photo-z codes being handled by a class that inherits qp.Ensemble features. I started the demo notebook in parallel with the metrics implementation. Now I am starting to make the QQ plots using DC1 functions adapted. I should push a couple of commits by the end of the day.

**Comment by gschwend** [comment]
Progress report: today I showed some progress in a demo notebook, where we have already the main metrics implemented (PIT outlier rate, CDE, KS, CvM, and AD). One issue I am dealing with is that the  KS, CvM, and AD metrics in DC1 code used the skgof library, which is not compatible with the scipy version required by qp. For the time being, I took a shortcut and changed the source code of skgof in my computer, so I don't get stuck. Now I'll work on rewriting the metrics function without using skgov based on Alex's suggestion (take advantage of qp CDF method to avoid having to do it all from scratch).

**Comment by gschwend** [comment]
Progress report: I replaced the statistical tests with scipy.stats functions in order to remove the dependency on skgov library. I also started a script inside the examples directory to run all metrics computation at once via the command line (besides the demo notebook where I show the usage of individual metrics separately).

**Comment by gschwend** [comment]
Progress report: it is not possible to just reuse scipy functions to compute all metrics because some of them do not allow the comparison with a uniform distribution, which is our goal. I am following Alex's suggestion of using qp.Ensemble CDF method applied to the PIT distribution, interpreting the PIT distribution as an ensemble of PDF containing only one PDF. I got PIT outlier rate and CDE Loss very close to DC1 paper results, but I still can't reproduce KS, CvM, and AD. So I am willing to open a pull request to ask @sschmidt23 to take a look at the code. For the time being, I have implemented the point estimate metrics (although it is out of the scope), as a second test, just to make sure I'm not making any mistake when reading the samples and computing the PITs. The results match! I reproduced all the point estimate metrics, so the attention can be focused just on the KS, CvM, and AD metrics. 
ps: I've just done a git rebase in my branch.

**Comment by gschwend** [comment]
Progress report: all metrics done, now under the subclass structure as it was in the original plan suggested by Alex. I removed completely the class "Sample". The demo and dc1 notebooks were updated. All metrics match with DC1 result, except for AD. The closest result is obtained when using scipy.stats.anderson_ksamp. The code is finished, but there are still minor updates to do in the classes' documentation. I think next week I'll finally start the unittest phase.
